### PR TITLE
VALID flag of Form's validation results callback is not accurate on initial mount

### PR DESCRIPTION
**Pull request by ypinneka**
#### What does this PR do?
Validation when run at component mount must keep track of required fields. Valid flag is not accurate at the initial mount time as validation is run before required fields are set. Adding a dependency of required fields array in the useEffect.

#### Where should the reviewer start?
Form.js

#### What testing has been done on this PR?
Verified the valid flag manually via stories

#### How should this be manually tested?
Make sure all the validations pass & leave a required field empty, the valid flag should stay false.

#### What are the relevant issues?
https://github.com/grommet/grommet/issues/5194

#### Do the grommet docs need to be updated?
no

#### Should this PR be mentioned in the release notes?
no

#### Is this change backwards compatible or is it a breaking change?
not a breaking change

### CODE REVIEW

**Comment by ypinneka** [comment]
Note: Right now, validation runs more than once when the component is mounted as there is a dependency on Required Fields array.

**Comment by ericsoderberghp** [comment]
> Note: Right now in cases where required prop is present, validation runs more than once when the component is mounted as there is a dependency on Required Fields array.

I think this illustrates a problem with the current approach. I am thinking we should convert `requiredFields` to use useRef, similar to how `validations` is managed.

I also noticed some eslint warnings in the current Form.js that should be addressed as well.

**Comment by ericsoderberghp** [comment]
Here's a revised version of Form.js that addresses some of the comments above and passes the tests:
```
import React, {
  forwardRef,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { FormContext } from './FormContext';

const defaultMessages = {
  invalid: 'invalid',
  required: 'required',
};
const defaultValue = {};
const defaultTouched = {};
const defaultValidationResults = {
  errors: {},
  infos: {},
};

// validations is an array from Object.entries()
const validate = (validations, value, omitValid) => {
  const nextErrors = {};
  const nextInfos = {};
  validations.forEach(([name, validation]) => {
    if (!omitValid) {
      nextErrors[name] = undefined;
      nextInfos[name] = undefined;
    }
    const result = validation(value[name], value);
    // typeof error === 'object' is implied for both cases of error with
    // a status message and for an error object that is a react node
    if (typeof result === 'object') {
      if (result.status === 'info') {
        nextInfos[name] = result.message;
      } else {
        nextErrors[name] = result.message || result; // could be a node
      }
    } else if (typeof result === 'string') {
      nextErrors[name] = result;
    }
  });
  return [nextErrors, nextInfos];
};

const Form = forwardRef(
  (
    {
      children,
      errors: errorsProp = defaultValidationResults.errors,
      infos: infosProp = defaultValidationResults.infos,
      messages = defaultMessages,
      onChange,
      onReset,
      onSubmit,
      onValidate,
      validate: validateOn = 'submit',
      value: valueProp,
      ...rest
    },
    ref,
  ) => {
    const [valueState, setValueState] = useState(valueProp || defaultValue);
    const value = useMemo(() => valueProp || valueState, [
      valueProp,
      valueState,
    ]);
    const [touched, setTouched] = useState(defaultTouched);
    const [validationResults, setValidationResults] = useState(
      defaultValidationResults,
    );

    // when onBlur input validation is triggered, we need to complete any
    // potential click events before running the onBlur validation.
    // otherwise, click events like reset, etc. may not be registered.
    // for a detailed scenario/discussion,
    // see: https://github.com/grommet/grommet/issues/4863
    // the value of pendingValidation is the name of the FormField
    // awaiting validation.
    const [pendingValidation, setPendingValidation] = useState(undefined);

    useEffect(() => {
      setPendingValidation(undefined);
      setValidationResults({ errors: errorsProp, infos: infosProp });
    }, [errorsProp, infosProp]);

    const validations = useRef({});
    const requiredFields = useRef([]);

    const buildValid = useCallback(
      nextErrors => {
        let valid = false;

        valid = requiredFields.current
          .filter(n => Object.keys(validations.current).includes(n))
          .every(
            field =>
              value[field] && (value[field] !== '' || value[field] !== false),
          );

        if (Object.keys(nextErrors).length > 0) valid = false;
        return valid;
      },
      [value],
    );

    // On initial mount, when validateOn is change or blur,
    // set validation results for any set fields and calculate whether
    // the form is valid overall.
    useEffect(() => {
      const validationsForSetFields = Object.entries(
        validations.current,
      ).filter(([n]) => value[n]);

      if (validationsForSetFields.length > 0 && validateOn !== 'submit') {
        const [errors, infos] = validate(validationsForSetFields, value);

        const nextValidationResults = {
          errors,
          infos,
          valid: buildValid(errors),
        };
        if (onValidate) onValidate(nextValidationResults);
        setValidationResults(nextValidationResults);
      }
      // We only want to run this for the value we have on initial mount.
      // We don't want subsequent changes to the value to re-run this.
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    // Currently, onBlur validation will trigger after a timeout of 120ms.
    useEffect(() => {
      const timer = setTimeout(() => {
        if (pendingValidation) {
          // run validations on the pending one and any other touched fields
          const [validatedErrors, validatedInfos] = validate(
            Object.entries(validations.current).filter(
              ([n]) => touched[n] || pendingValidation.includes(n),
            ),
            value,
          );
          setPendingValidation(undefined);

          setValidationResults(prevValidationResults => {
            // keep any previous errors and infos for untouched keys,
            // these may have come from a submit
            const nextErrors = {
              ...prevValidationResults.errors,
              ...validatedErrors,
            };
            const nextInfos = {
              ...prevValidationResults.infos,
              ...validatedInfos,
            };

            // Remove any errors or infos that we don't have any validations
            // for anymore. This can occur when fields are dynamically removed.
            Object.keys(nextErrors)
              .filter(
                n => !validations.current[n] || nextErrors[n] === undefined,
              )
              .forEach(n => delete nextErrors[n]);
            Object.keys(nextInfos)
              .filter(
                n => !validations.current[n] || nextInfos[n] === undefined,
              )
              .forEach(n => delete nextInfos[n]);

            const nextValidationResults = {
              errors: nextErrors,
              infos: nextInfos,
              valid: buildValid(nextErrors),
            };
            if (onValidate) onValidate(nextValidationResults);
            return nextValidationResults;
          });
        }
        // a timeout is needed to ensure that a click event (like one on a reset
        // button) completes prior to running the validation. without a timeout,
        // the blur will always complete and trigger a validation prematurely
        // The following values have been empirically tested, but 120 was
        // selected because it is the largest value
        // Chrome: 100, Safari: 120, Firefox: 80
      }, 120);

      return () => clearTimeout(timer);
    }, [
      buildValid,
      pendingValidation,
      onValidate,
      touched,
      value,
      requiredFields,
    ]);

    // clear any errors when value changes
    useEffect(() => {
      if (validateOn !== 'change') setPendingValidation(undefined);
      setValidationResults(prevValidationResults => {
        const [nextErrors, nextInfos] = validate(
          Object.entries(validations.current).filter(
            ([n]) =>
              prevValidationResults.errors[n] || prevValidationResults.infos[n],
          ),
          value,
        );
        return {
          errors: { ...prevValidationResults.errors, ...nextErrors },
          infos: { ...prevValidationResults.infos, ...nextInfos },
        };
      });
    }, [touched, validateOn, value]);

    // There are three basic patterns of handling form input value state:
    //
    // 1 - form controlled
    //
    // In this model, the caller sets `value` and `onChange` properties
    // on the Form component to supply the values used by the input fields.
    // In useFormContext(), componentValue would be undefined and formValue
    // is be set to whatever the form state has. Whenever the form state
    // changes, we update the contextValue so the input component will use
    // that. When the input component changes, we will call update() to
    // update the form state.
    //
    // 2 - input controlled
    //
    // In this model, the caller sets `value` and `onChange` properties
    // on the input components, like TextInput, to supply the value for it.
    // In useFormContext(), componentValue is this value and we ensure to
    // update the form state, via update(), and set the contextValue from
    // the componentValue. When the input component changes, we will
    // call update() to update the form state.
    //
    // 3 - uncontrolled
    //
    // In this model, the caller doesn't set a `value` or `onChange` property
    // at either the form or input component levels.
    // In useFormContext(), componentValue is undefined and valueProp is
    // undefined and nothing much happens here. That is, unless the
    // calling component needs to know the state in order to work, such
    // as CheckBox or Select. In this case, those components supply
    // an initialValue, which will trigger updating the contextValue so
    // they can have access to it.
    //
    const useFormInput = (name, componentValue, initialValue) => {
      const [inputValue, setInputValue] = useState(initialValue);
      const formValue = name ? value[name] : undefined;

      // This effect is for pattern #2, where the controlled input
      // component is driving the value via componentValue.
      useEffect(() => {
        if (
          name && // we have somewhere to put this
          componentValue !== undefined && // input driving
          componentValue !== formValue // don't already have it
        ) {
          setValueState(prevValue => {
            const nextValue = { ...prevValue };
            nextValue[name] = componentValue;
            return nextValue;
          });
          // don't onChange on programmatic changes
        }
      }, [componentValue, formValue, name]);

      let useValue;
      if (componentValue !== undefined)
        // input component drives, pattern #2
        useValue = componentValue;
      else if (valueProp && name && formValue !== undefined)
        // form drives, pattern #1
        useValue = formValue;
      else if (formValue === undefined && name)
        // form has reset, so reset input value as well
        useValue = initialValue;
      else useValue = inputValue;

      return [
        useValue,
        nextComponentValue => {
          if (name) {
            // we have somewhere to put this
            const nextTouched = { ...touched };
            nextTouched[name] = true;

            if (!touched[name]) {
              // don't update if not needed
              setTouched(nextTouched);
            }

            const nextValue = { ...value };
            nextValue[name] = nextComponentValue;
            setValueState(nextValue);
            if (onChange) onChange(nextValue, { touched: nextTouched });
          }
          if (initialValue !== undefined) setInputValue(nextComponentValue);
        },
      ];
    };

    const useFormField = ({
      error: errorArg,
      info: infoArg,
      name,
      required,
      validate: validateArg,
    }) => {
      const error = errorArg || validationResults.errors[name];
      const info = infoArg || validationResults.infos[name];

      useEffect(() => {
        const validateSingle = (aValidate, value2, data) => {
          let result;
          if (typeof aValidate === 'function') {
            result = aValidate(value2, data);
          } else if (aValidate.regexp) {
            if (!aValidate.regexp.test(value2)) {
              result = aValidate.message || messages.invalid;
              if (aValidate.status) {
                result = { message: result, status: aValidate.status };
              }
            }
          }
          return result;
        };

        const validateField = (value2, data) => {
          let result;
          if (
            required &&
            // false is for CheckBox
            (value2 === undefined || value2 === '' || value2 === false)
          ) {
            result = messages.required;
          } else if (validateArg) {
            if (Array.isArray(validateArg)) {
              validateArg.some(aValidate => {
                result = validateSingle(aValidate, value2, data);
                return !!result;
              });
            } else {
              result = validateSingle(validateArg, value2, data);
            }
          }
          return result;
        };

        const index = requiredFields.current.indexOf(name);
        if (required) {
          if (index === -1) requiredFields.current.push(name);
        } else if (index !== -1) requiredFields.current.splice(index, 1);

        if (validateArg || required) {
          validations.current[name] = validateField;
          return () => delete validations.current[name];
        }

        return undefined;
      }, [error, name, required, validateArg]);

      return {
        error,
        info,
        inForm: true,
        onBlur:
          validateOn === 'blur'
            ? () =>
                setPendingValidation(
                  pendingValidation ? [...pendingValidation, name] : [name],
                )
            : undefined,
        onChange:
          validateOn === 'change'
            ? () =>
                setPendingValidation(
                  pendingValidation ? [...pendingValidation, name] : [name],
                )
            : undefined,
      };
    };

    return (
      <form
        ref={ref}
        {...rest}
        onReset={event => {
          setPendingValidation(undefined);
          if (!valueProp) {
            setValueState(defaultValue);
            if (onChange) onChange(defaultValue, { touched: defaultTouched });
          }
          setTouched(defaultTouched);
          setValidationResults(defaultValidationResults);

          if (onReset) {
            event.persist(); // extract from React's synthetic event pool
            const adjustedEvent = event;
            adjustedEvent.value = defaultValue;
            onReset(adjustedEvent);
          }
        }}
        onSubmit={event => {
          // Don't submit the form via browser form action. We don't want it
          // if the validation fails. And, we assume a javascript action handler
          // otherwise.
          event.preventDefault();
          setPendingValidation(undefined);
          const [nextErrors, nextInfos] = validate(
            Object.entries(validations.current),
            value,
            true,
          );

          setValidationResults(() => {
            const nextValidationResults = {
              errors: nextErrors,
              infos: nextInfos,
              // Show form's validity when clicking on Submit
              valid: buildValid(nextErrors),
            };
            if (onValidate) onValidate(nextValidationResults);
            return nextValidationResults;
          });

          if (Object.keys(nextErrors).length === 0 && onSubmit) {
            event.persist(); // extract from React's synthetic event pool
            const adjustedEvent = event;
            adjustedEvent.value = value;
            adjustedEvent.touched = touched;
            onSubmit(adjustedEvent);
          }
        }}
      >
        <FormContext.Provider value={{ useFormField, useFormInput }}>
          {children}
        </FormContext.Provider>
      </form>
    );
  },
);

Form.displayName = 'Form';

let FormDoc;
if (process.env.NODE_ENV !== 'production') {
  FormDoc = require('./doc').doc(Form); // eslint-disable-line global-require
}

const FormWrapper = FormDoc || Form;

export { FormWrapper as Form };
```

**Comment by ypinneka** [comment]
Screen Shot 2021-04-28 at 8 48 58 AM

The above code would cause a regression with the error/info format, attaching a screenshot when there are no errors/infos, the object has undefined & this is only when validation is run on mount.
To counter this I have added the same filter you have in place for validations on mount.

**Comment by halocline** [comment]
> I also noticed some eslint warnings in the current Form.js that should be addressed as well.

Just now seeing this thread and PR. [Related comment](https://github.com/grommet/grommet/pull/5187/files#r622514773).

**Comment by ypinneka** [comment]
> > I also noticed some eslint warnings in the current Form.js that should be addressed as well.
> 
> Just now seeing this thread and PR. [Related comment](https://github.com/grommet/grommet/pull/5187/files#r622514773).

They have been addressed here.

**Comment by halocline** [comment]
> > I also noticed some eslint warnings in the current Form.js that should be addressed as well.
> 
> Just now seeing this thread and PR. [Related comment](https://github.com/grommet/grommet/pull/5187/files#r622514773).

It looks like changes here address the items I was running into. Reviewing Storybook stories now.

File: src/js/components/Form/Form.js [hunk 1]
```diff
@@ -110,40 +121,31 @@ const Form = forwardRef(
         .forEach(n => delete nextInfos[n]);
     };
 
+    // On initial mount, when validateOn is change or blur,
+    // set validation results for any set fields and calculate whether
+    // the form is valid overall.
```
Comment (halocline) [inline src/js/components/Form/Form.js:126]: Thank you for adding.

File: src/js/components/Form/Form.js [hunk 2]
```diff
@@ -110,40 +121,31 @@ const Form = forwardRef(
...
+          errors,
+          infos,
+          valid: buildValid(errors),
+        };
+        if (onValidate) onValidate(nextValidationResults);
+        setValidationResults(nextValidationResults);
       }
+      // We only want to run this for the value we have on initial mount.
+      // We don't want subsequent changes to the value to re-run this.
+      // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (halocline) [inline src/js/components/Form/Form.js:148]: Thank you for adding.

**Review by halocline — APPROVED** [review]
This addresses the issues I was seeing. Thank you.

**Review by taysea — APPROVED** [review]
LGTM. I checked the storybook deploy and all seemed to be behaving as expected across Form examples.

**Review by ericsoderberghp — APPROVED** [review]

**Pull request merged**